orion/primitives/tadgan.py

```diff
@@ -3,77 +3,141 @@
+        validation_split (Optional[Float]): Number between 0 and 1. Fraction of the training data
+            to be used as validation data. Default 0.2.
+        callbacks (Optional[dict]): Callbacks to apply during training.
+        verbose (Optional[Union[int, bool]]): Verbosity mode where 0 = silent, 1 = progress bar,
+            2 = one line per epoch. Default False.
+        detailed_losses (Optional[bool]): Whether to output all loss values in verbose mode.
+        **hyperparameters (Optional[dict]): additional inputs.
+    """
 
+    def __init__(self, layers_encoder: list, layers_generator: list, layers_critic_x: list,
+                 layers_critic_z: list, optimizer: str, input_shape: tuple = (100, 1),
+                 latent_shape: Optional[tuple] = None, target_shape: Optional[tuple] = None,
+                 latent_dim: int = 20, learning_rate: float = 0.0005, epochs: int = 2000,
+                 batch_size: int = 64, iterations_critic: int = 5, shuffle: bool = True,
+                 callbacks: tuple = tuple(), validation_ratio: float = 0.2,
+                 detailed_losses: bool = True, verbose: Union[int, bool] = True,
+                 **hyperparameters):
+        """Initialize the TadGAN model."""
+        super(TadGAN, self).__init__()
 
-class RandomWeightedAverage(_Merge):
-    def _merge_function(self, inputs):
-        """
-        Args:
-            inputs[0] x     original input
-            inputs[1] x_    predicted input
-        """
-        alpha = K.random_uniform((64, 1, 1))
-        return (alpha * inputs[0]) + ((1 - alpha) * inputs[1])
-
+        # Required model hyperparameters to construct model.
+        self.layers_encoder, self.layers_generator = layers_encoder, layers_generator
+        self.layers_critic_x, self.layers_critic_z = layers_critic_x, layers_critic_z
 
-class TadGAN(object):
-    """TadGAN model for time series reconstruction.
+        # Optional model hyperparameters.
+        self.shape = input_shape
+        self.latent_dim = latent_dim
+        self.latent_shape = latent_shape if latent_shape else (latent_dim, 1)
+        self.target_shape = target_shape if target_shape else input_shape
+
+        # Calculated model hyperparameters
+        self.encoder_input_shape = hyperparameters.get('encoder_input_shape', self.shape)
+        self.generator_input_shape = hyperparameters.get('generator_input_shape',
+                                                         self.latent_shape)
+        self.critic_x_input_shape = hyperparameters.get('critic_x_input_shape', self.target_shape)
+        self.critic_z_input_shape = hyperparameters.get('critic_z_input_shape', self.latent_shape)
+
+        self._setdefault(hyperparameters, 'generator_reshape_dim', self.shape[0] // 2)
+        self._setdefault(hyperparameters, 'generator_reshape_shape', (self.shape[0] // 2, 1))
+        self._setdefault(hyperparameters, 'encoder_reshape_shape', self.latent_shape)
```

Comment:
With respect to overriding defaults, there was an `_augment_hyperparameters` to "infer" the shape of the inputs and outputs of the encoder and generator networks (which include this block).

For example, let's assume the signal is of shape `[window_size, num_channels]` e.g. `[100, 1]` in most of our examples. If the primitive has `self.shape=input_shape=200` I want to edit that parameter and make `self.shape=100` and this was the purpose of `_augment_hyperparameters` and `_set_shapes` previously which were called in the `fit` function.

I noticed you did not include it in the refactored code. Is there a problem if we still try to include it to override these hyperparameters?
It might help if you refer to https://github.com/sintel-dev/Orion/pull/249

Reply:
Brought this back!